src/commands/medusaCodeLens.ts

```diff
@@ -15,11 +22,134 @@ export async function generateMedusaDatasetCommand(documentUri: Uri): Promise<vo
-  await window.showInformationMessage(
-    `Generate Medusa Dataset clicked for: ${documentUri?.fsPath || "Unknown file"}`,
-  );
+  try {
+    // Read the Avro schema file contents
+    const { content: avroSchemaContent } = await getEditorOrFileContents(documentUri);
+
+    if (!avroSchemaContent.trim()) {
+      await window.showErrorMessage("The Avro schema file is empty.");
+      return;
+    }
+
+    logger.info("Reading Avro schema file", {
+      filePath: documentUri.fsPath,
+      contentLength: avroSchemaContent.length,
+    });
+
+    // Show progress while calling the API
+    const dataset = await window.withProgress(
+      {
+        location: { viewId: "confluent" },
+        title: "Generating Medusa Dataset...",
+        cancellable: false,
+      },
+      () => convertAvroSchemaToDataset(avroSchemaContent),
+    );
+
+    // Save the dataset to a file
+    await saveDatasetToFile(dataset);
+  } catch (error) {
+    logger.error("Failed to generate Medusa dataset", error);
+    const errorMessage = error instanceof Error ? error.message : "Unknown error occurred";
+    await window.showErrorMessage(`Failed to generate Medusa dataset: ${errorMessage}`);
+  }
+}
+
+async function convertAvroSchemaToDataset(avroSchemaContent: string): Promise<DatasetDTO> {
+  // Parse and validate the Avro schema content
+  let parsedSchema;
+  try {
+    parsedSchema = JSON.parse(avroSchemaContent);
+  } catch (parseError) {
+    throw new Error(
+      `Invalid JSON in Avro schema file: ${parseError instanceof Error ? parseError.message : "Unknown parsing error"}`,
+    );
+  }
+
+  const medusaContainer = await getMedusaContainer();
+  if (!medusaContainer) {
+    throw new Error("Medusa container not found. Please start the local Medusa service.");
+  }
+  const medusaPort = getContainerPublicPort(medusaContainer);
+  if (!medusaPort) {
+    throw new Error("Medusa container port not accessible. Please check container configuration.");
+  }
+  logger.info("Calling Medusa API", {
+    port: medusaPort,
+    schemaType: parsedSchema.type,
+    schemaName: parsedSchema.name,
+  });
+
+  // Call the Medusa API to convert the Avro schema to dataset
+  const medusaApi = getMedusaSchemaManagementApi(medusaPort);
+  const dataset = await medusaApi.convertAvroSchemaToDataset({
+    body: parsedSchema,
+  });
+
+  logger.info("Successfully generated Medusa dataset", {
+    eventsCount: dataset.events?.length || 0,
+  });
+
+  return dataset;
+}
+
+/**
+ * Prompts the user to save a DatasetDTO to a file with .dataset.json extension.
+ * Defaults to the current workspace directory.
+ */
+async function saveDatasetToFile(dataset: DatasetDTO): Promise<void> {
+  try {
+    // Get the workspace folder as default location
+    const workspaceFolder = workspace.workspaceFolders?.[0];
+    const defaultUri = workspaceFolder ? workspaceFolder.uri : undefined;
+    const defaultFilename = `my-dataset.dataset.json`;
```

Comment:
Can we default the dataset filename off of the schema's subject if possible? Or is that too far long gone by now?